### ISSUE TITLE
Packet API

### ISSUE THREAD

**Issue by PizzaCrust** [FountainMC/FountainAPI#18]

**Comment by wgaylord** [comment]
This is a great idea. (I wanted to do this for Dimensions.)

**Comment by PizzaCrust** [comment]
Its a core feature if forge mods are supported to enable plugins to interact with mods via packets.

**Comment by Techcable** [comment]
I say we keep the packet API as a unsafe API in a special package and as a separate maven dependency if we're going to do it at all. Sort of like `sun.mic.Unsafe`

**Comment by PizzaCrust** [comment]
just put it into a **.internal.**\* package

**Comment by Techcable** [comment]
yes, and no reflection utilities or anything, just `PacketMagic.registerPacketListener(Consumer<Object>)` and `PacketMagic.sendPacket(Player, Object)`

**Comment by minecrafter** [comment]
**Oppose**, for portability's sake. Fountain can provide a few things to help, but in general I feel this is best handled outside Fountain.

**Comment by wgaylord** [comment]
We need to have it on foutain other wise people will be digging around with asm and reflection which is not good.

**Comment by phase** [comment]
Putting in the API is a bad idea. We could have it in the the implementation, so if people want to build off a specific version they get that version's Packets.

**Comment by Techcable** [comment]
Unless we're going to wrap all packets in a version-safe manner, plugins will have to access packet fields with reflection.
Therefore we should put the `PacketMagic` class in the implementation, and then plugins like ProtocolLib can provide a safer and more advanced wrapper, but will have to hook into our internals to do so.

**Comment by PizzaCrust** [comment]
Sponge has this already implemented. (**Custom Payload Packet**), https://github.com/SpongePowered/SpongeAPI/pull/363/commits/e9f96b70c943d267de6756fe91dcc180a843fcdf

**Comment by minecrafter** [comment]
Nope. That Sponge PR refers to plugin messaging.

**Comment by PizzaCrust** [comment]
> Add ChannelBuf. ChannelBuf is an abstract interface
> to represent a ByteBuffer to allow plugins and the server
> to read and write data without depending on a specific
> networking library.

That are packets. Packets =  Messages.

**Comment by minecrafter** [comment]
I have worked with this specific Sponge API. This is the standard Plugin Message packet in Minecraft. Fountain will support listening and sending messages with this specific packet.

**Comment by PizzaCrust** [comment]
Oh, good.
